Library Evolution for Stable ABIs (#1025)
apple/swift-evolution: proposals/NNNN-library-evolution.md

  [SE-0192]: https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md

## Proposed solution

The compiler will gain a new mode, called "library evolution mode". This mode will be off by default. A library compiled with this mode enabled is said to be _ABI-stable_.

When a library is compiled with library evolution mode enabled, it is not an ABI-breaking change to modify the fields in a struct (to add, remove, or reorder them), and to add new enum cases (including with associated values). This implies that clients must manipulate fields and enum cases indirectly, via non-inlinable function calls. Information such as the size of the type, and the layout of its fields, becomes something that can only be determined at runtime. Types that reserve this flexibility are referred to as "resilient types".

> A new `@frozen` keyword will be introduced to allow library authors to opt out of this flexibility on a per-type basis. This keyword promises that stored instance properties within the struct will not be *added,* *removed,* or *reordered*, and that an enum will never *add,* *remove,* or *reorder* its cases (note removing, and sometimes reordering, cases can already be source breaking, not just ABI breaking). The compiler will use this for optimization purposes when compiling clients of the type. The precise set of allowed changes is defined below.

Reviewer: `@frozen` an attribute, not a keyword, correct?

Author: yes! thanks.

Author: > Okay, as an implicit contrast with the column to the left?

Yes. There are some things that affect the ABI even without frozen, which aren't bolded.

> I think it might be better to be stronger about this than "Affects".

Yeah, I guess "Not Allowed". The question would be, not allowed by whom, given the compiler isn't going to stop you. @jrose-apple do you have an opinion?

Reviewer 2: (Somehow this ended up attached to the wrong comment)

"Not allowed" works for me.  "Allowed" has the saw actorship problem. :)

A new command-line argument, `-enable-library-evolution`, will enable this new mode.

Turning on library evolution mode will have the following effects on source code:

- The default behavior for enums will change to be non-frozen. This is the only change visible to _users_ of the library, who will need to use the `@unknown default:` technique described in SE-0192.
- To ensure that all fields are initialized, inlinable `init`s of the type must delegate to a non-inlinable `init`.

### `@frozen` on `struct` types

When a library author is certain that there will never be a need to add fields to a struct in future, they may mark that type as `@frozen`. 

This will allow the compiler to optimize away at compile time some calls that would otherwise need to be made at runtime (for example, it might access fields directly without the indirection).

When compiling with binary stability mode on, a struct can be marked `@frozen` as long as it meets all of the following conditions:

- The struct is *ABI-public* (see [SE-0193][]), i.e. `public` or marked `@usableFromInline`.
- Every class, enum, struct, protocol, or typealias mentioned in the types of the struct's fields is ABI-public.

> - No fields have observing accessors (`willSet` or `didSet`).

Reviewer 3: Maybe the forum is the best place to discuss this but I feel like this restriction is somewhat arbitrary and doesn’t buy us much.

- If a field has an initial value, the expression computing the initial value does not reference any types or functions that are not ABI-public.

```swift
@frozen
public struct Point {
  public var x: Double
  public var y: Double

  public init(_ x: Double, _ y: Double) {
    self.x = x
    self.y = y
  }
}
```

This affects what changes to the struct's fields affect the ABI of the containing library:

| Change | Normal struct | `@frozen` struct
|---|:---:|:---:
| Adding fields | Allowed | **Affects ABI**
| Reordering fields | Allowed | **Affects ABI**
| Removing ABI-public fields | Affects ABI | Affects ABI
| Removing non-ABI-public fields | Allowed | **Affects ABI**
| Changing the type of an ABI-public field | Affects ABI | Affects ABI
| Changing the type of a non-ABI-public field | Allowed | **Affects ABI**
| Changing a stored instance property to computed | Allowed | **Affects ABI**
| Changing a computed instance property to stored | Allowed | **Affects ABI**
| Changing the access of a non-ABI-public field | Allowed | Allowed

> | Marking an `internal` field as `@abiPublic` | Allowed | Allowed

Reviewer 3: Typo: it’s @usableFroInline not @abiPublic

# Library Evolution for Stable ABIs

* Proposal: [SE-NNNN](NNNN-library-evolution.md)
* Authors: [Jordan Rose](https://github.com/jrose-apple), [Ben Cohen](https://github.com/airspeedswift)
* Review Manager: TBD
* Status: **Awaiting review**

> * Implementation: Implemented in Swift 5 for standard library use

Reviewer 2: Could you whip up an implementation PR that actually uses the command-line and attribute spelling you're proposing?

Author: PR here, but it's far easier to use the shipping toolchain to try it out with the underscores than one generated from that PR.

Reviewer 2: PR where?

Author: Link added.

Reviewer 2: Ah, I see it, thanks.

- No fields have observing accessors (`willSet` or `didSet`).
- If a field has an initial value, the expression computing the initial value does not reference any types or functions that are not ABI-public.

```swift
@frozen
public struct Point {
  public var x: Double
  public var y: Double

  public init(_ x: Double, _ y: Double) {
    self.x = x
    self.y = y
  }
}
```

This affects what changes to the struct's fields affect the ABI of the containing library:

| Change | Normal struct | `@frozen` struct
|---|:---:|:---:
| Adding fields | Allowed | **Affects ABI**
| Reordering fields | Allowed | **Affects ABI**
| Removing ABI-public fields | Affects ABI | Affects ABI
| Removing non-ABI-public fields | Allowed | **Affects ABI**
| Changing the type of an ABI-public field | Affects ABI | Affects ABI
| Changing the type of a non-ABI-public field | Allowed | **Affects ABI**
| Changing a stored instance property to computed | Allowed | **Affects ABI**
| Changing a computed instance property to stored | Allowed | **Affects ABI**
| Changing the access of a non-ABI-public field | Allowed | Allowed
| Marking an `internal` field as `@usableFromInline` | Allowed | Allowed

> | Changing an `internal` ABI-public field to be `public` | Allowed | Allowed

Reviewer 2: What's with the inconsistent bolding?

Author: Bold indicates inconsistency between frozen and unfrozen.

Reviewer 2: And how am I as a reader supposed to know that?

2. The representation of `nil` for any type that is "nullable" in C / Objective-C is the same as the representation of `nil` or `NULL` in those languages. This includes class references, class-bound protocol references, class type references, unsafe-pointers, `@convention(c)` functions, `@convention(block)` functions, OpaquePointer, Selector, and NSZone. This has been true since Swift 3 ([SE-0055][]).

This proposal does not change either of these guarantees.

  [trivial]: https://docs.microsoft.com/en-us/cpp/cpp/trivial-standard-layout-and-pod-types
  [SE-0055]: https://github.com/apple/swift-evolution/blob/master/proposals/0055-optional-unsafe-pointers.md

### `@frozen` on `enum` types

Marking an enum as `@frozen` will similarly allow the compiler to optimize away runtime calls.

In addition, marking an enum as frozen restores the ability of a library user to exhaustively switch over that enum without an `@unknown default:`, because it guarantees no further cases will be added.

Once frozen, all changes made to an enum's cases affect its ABI.

## Naming

> `@frozen` was used here, as originally used in [SE-0192][], for both fixed-layout structs and enums. 

Reviewer 2: Please fill in these links and make any other editorial changes you intend to make before this goes up.

Author: The link's there, it's (implicit) reference style markdown link. Renders correctly in GitHub.

A new `@frozen` keyword will be introduced to allow library authors to opt out of this flexibility on a per-type basis. This keyword promises that stored instance properties within the struct will not be *added,* *removed,* or *reordered*, and that an enum will never *add,* *remove,* or *reorder* its cases (note removing, and sometimes reordering, cases can already be source breaking, not just ABI breaking). The compiler will use this for optimization purposes when compiling clients of the type. The precise set of allowed changes is defined below.

This attribute has no effect when library evolution mode is off. In that case, the compiler acts as if *every* struct or enum in the library being compiled is "frozen".

## Detailed design

### Library evolution mode

A new command-line argument, `-enable-library-evolution`, will enable this new mode.

> Turning on library evolution mode will have the following affects on source code:

Reviewer 4: ```suggestion
Turning on library evolution mode will have the following effects on source code:
```

- The default behavior for enums will change to be non-frozen. This is the only change visible to _users_ of the library, who will need to use the `@unknown default:` technique described in SE-0192.
- To ensure that all fields are initialized, inlinable `init`s of the type must delegate to a non-inlinable `init`.

### `@frozen` on `struct` types

When a library author is certain that there will never be a need to add fields to a struct in future, they may mark that type as `@frozen`. 

> This will allow the compiler to optimize away at compile time some calls that would otherwise need to be made at runtime (for example, it might access ).

Reviewer 4: Example incomplete, what might it access?

## Source compatibility

This change does not impact source compatibility. It has always been a source-compatible change to modify a struct's non-public stored instance properties, and to add a new public stored instance property, as long as the struct's existing API does not change (including any initializers).

## Effect on ABI stability

Currently, the layout of a public struct is known at compile time in both the defining library and in its clients. For a library concerned about binary compatibility, the layout of a non-fixed-contents struct must not be exposed to clients, since the library may choose to add new stored instance properties that do not fit in that layout in its next release, or even remove existing properties as long as they are not public.

These considerations should not affect libraries shipped with their clients, including SwiftPM packages. These libraries should always have library evolution mode turned off, indicating that the compiler is free to optimize based on the layout of a type (because the library won't change).

## Effect on Library Evolution

> Both structs and enums can gain new protocol conformances or methods, even when `@frozen`. Binary compatability only affects additions of fields or cases.

Reviewer 4: ```suggestion
Both structs and enums can gain new protocol conformances or methods, even when `@frozen`. Binary compatibility only affects additions of fields or cases.
```

If we wanted to add this later, the way we would probably implement it is by saying that accessors for stored instance properties of `@frozen` structs must be marked `@inlinable`, with all of the implications that has for being able to make changes in the future.

## Alternatives considered

### Annotation syntax

#### Naming

`@frozen` comes from [SE-0192][], where it was originally used as a term for for enums that will not gain new cases.

Other than `@frozen`, most names have revolved around the notion of "fixed" somehow: `@fixedContents`, `@fixedLayout`, or even just `@fixed`. We ultimately chose `@frozen` as there is a strong similarity between this feature for both enums and structs.

`final` was suggested at one point, but the meaning is a little far from that of `final` on classes.

#### Modifier or attribute?

> This proposal suggests a new *attribute* for structs, `@fixedContents`; it could also be a modifier `fixedContents` (or `fixed`), implemented as a context-sensitive keyword. Because this annotation only affects the struct's *implementation* rather than anything about its use, an attribute seemed more appropriate.

Reviewer 4: ```suggestion
This proposal suggests a new *attribute* for structs, `@frozen`; it could also be a modifier `frozen`, implemented as a context-sensitive keyword. Because this annotation only affects the struct's *implementation* rather than anything about its use, an attribute seemed more appropriate.
```

Author: Ah, needed to be generilized for enums too.

### "No reordering" vs. "No renaming"

As written, this proposal disallows *reordering* field of a fixed-contents struct. This matches the restrictions of C structs, but is different from everything else in Swift, which freely allows reordering. Why?

Let's step back and remember our goal: a stable, unchanging ABI for a fixed-contents struct. That means that whatever changes are made to the struct, the in-memory representation and the rest of the ABI must be consistent across library versions. We can think of this as a function that takes fields and produces a layout:

```swift
struct Field {
  var name: String
  var type: ValueType
}
typealias Offset = Int
func layoutForFixedContentsStruct(fields: [Field]) -> [Offset: Field]
```

To make things simple, let's assume all the fields in the struct have the same type, like this PolarPoint:

```swift

> @fixedContents public struct PolarPoint {

Reviewer 4: ```suggestion
@frozen public struct PolarPoint {
```

  // You probably wouldn't do that in real life,
  // but it's needed for this contrived example.
  public var radius: Double
  private var angle: Double

  public init(x: Double, y: Double) {
    self.radius = sqrt(x * x, y * y)
    self.angle = atan2(y, x)
  }

  public var x: Double {
    return radius * cos(angle)
  }
  public var y: Double {
    return radius * sin(angle)
  }
}
```

We have two choices for how to lay out the PolarPoint struct:

- `[0: radius, 8: angle]`
- `[0: angle, 8: radius]`

*It's important that everyone using the struct agrees on which layout we're using.* Otherwise, an angle could get misinterpreted as a radius when someone outside the module tries to access the `radius` property.

So how do we decide? We have three choices:

1. Sort the fields by declaration order.
2. Sort the fields by name.
3. Do something clever with access levels.

Which translates to three possible restrictions:

1. Fields of a fixed-contents struct may not be reordered, even non-public ones.
2. Fields of a fixed-contents struct may not be renamed, even non-public ones.
3. Copying the struct requires calling a function because we can't see the private fields.

> (3) is a non-starter, since it defeats the performance motivation for `@fixedContents`. That leaves (1) and (2), and we decided it would be *really weird* if renaming the private property `angle` to `theta` changed the layout of the struct. So the restriction on reordering fields was considered the least bad solution. It helps that it matches the behavior of C, Rust, and other languages.

Reviewer 4: ```suggestion
(3) is a non-starter, since it defeats the performance motivation for `@frozen`. That leaves (1) and (2), and we decided it would be *really weird* if renaming the private property `angle` to `theta` changed the layout of the struct. So the restriction on reordering fields was considered the least bad solution. It helps that it matches the behavior of C, Rust, and other languages.
```

# Library Evolution for Stable ABIs

> * Proposal: [SE-NNNN](NNNN-non-exhaustive-enums.md)

Reviewer 5: Needs basename fix, but I guess that's okay.

Author: oops

> * Authors: [Jordan Rose](https://github.com/jrose-apple)

Reviewer 5: Give yourself some credit. :-)

This proposal introduces:

- a "library evolution" build mode for libraries that are declaring ABI stability, which preserves the ability to make certain changes to types without breaking the library's ABI; and
- an attribute for such libraries to opt out of this flexibility on a per-type basis, allowing certain compile-time optimizations.

The mechanisms for this are already in place, and were used to stabilize the
ABI of the standard library. This proposal makes them features for use by
any 3rd-party library that wishes to declare itself ABI stable.

**This build mode will have no impact on libraries built and distributed with an app**. Such libraries will receive the same optimization that they have in previous versions of Swift.

_Note: this proposal will use the word "field" to mean "stored instance property"._

## Motivation

> Starting in Swift 5, libraries are able to declare a stable ABI, allowing a  library binary to be replaced with a newer version without requiring client programs to be recompiled. 

Reviewer 5: "Starting in Swift 5" might not really make sense now.

Author: "as of" I guess

Reviewer 5: Ah, I meant that it's something we did for the standard library and overlays, but that isn't generally available.

This often leads to manual workarounds. A common technique is to have the struct hold only a pointer to an "impl" type, which holds the actual stored properties. All access to these properties is made via function calls, which can be updated to handle changes to the layout of the "impl" type. This has some obvious downsides:

- it introduces indirection and function call overhead when accessing fields;
- it means the fields of the struct must be stored on the heap even when they could otherwise be stored on the stack, with that heap allocation needing to be managed via reference counting or some other mechanism;
- it means that the fields of the struct data are not stored in contiguous memory when placed in an array; and 
- the implementation of all this must be provided by the library author.

> A similar challenge occurs with enums. As discussed in [SE-0192][], introducing new cases to an enum can break source compatability. There are also ABI consequences to this: adding new enum cases sometimes means increasing the storage needed for the enum, which can affect the size and layout of an enum in the same way adding fields to a struct can.

Reviewer 5: Typo: "compatability". Also, you may want to distinguish the C and Swift cases here. In C (well, Objective-C) this isn't a problem because you specify an underlying type.

The goal of this proposal is to reduce the burden on library developers by building into the compiler an automatic mechanism to reserve the flexibility to alter the internal representation of structs and enums, without manual workarounds. This mechanism can be implemented in such a way that optimizations such as stack allocation or contiguous inline storage of structs can still happen, while leaving the size of the type to be determined at runtime. 

  [SE-0192]: https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md

## Proposed solution

The compiler will gain a new mode, called "library evolution mode". This mode will be off by default. A library compiled with this mode enabled is said to be _ABI-stable_.

> When a library is compiled with library evolution mode enabled, it is not an ABI-breaking change to modify the fields in a struct (to add, remove, or reorder them), and to add new enum cases (including with associated values). This implies that clients must manipulate fields and enum cases indirectly, via non-inlable function calls. Information such as the size of the type, and the layout of its fields, becomes something that can only be determined at runtime. Types that reserve this flexibility are referred to as "resilient types".

Reviewer 5: Typo: "inlable". Also, I think we still want to get rid of "resilient".

Author: I think there's benefit from having an official term of art for a type you can still muck with the internals of.

Reviewer 5: I think so too, but I don't think it's "resilient". It's not a "term of art" outside of Swift.

> A new `@frozen` keyword will be introduced to allow library authors to opt out of this flexibility on a per-type basis. This keyword promises that stored instance properties within the struct will not be *added,* *removed,* or *reordered*. The compiler will use this for optimization purposes when compiling clients of the struct. The precise set of allowed changes is defined below.

Reviewer 5: Enums should be mentioned in this paragraph too.

> This attribute has no effect when binary stability mode is off. In that case, the compiler acts as if *every* struct or enum in the library being compiled is "frozen".

Reviewer 5: Lingering reference to "binary stability mode"

## Detailed design

### Library evolution mode

A new command-line argument, `-enable-library-evolution`, will enable this new mode.

Turning on library evolution mode will have the following affects on source code:

- The default behavior for enums will change to be non-frozen. This is the only change visible to _users_ of the library, who will need to use the `@unknown default:` technique described in SE-0192.
- To ensure that all fields are initialized, inlinable `init`s of the type must delegate to a non-inlinable `init`.

### `@frozen` on `struct` types

When a library author is certain that there will never be a need to add fields to a struct in future, they may mark that type as `@frozen`. 

This will allow the compiler to optimize away at compile time some calls that would otherwise need to be made at runtime (for example, it might access ).

When compiling with binary stability mode on, a struct can be marked `@frozen` as long as it meets all of the following conditions:

> - The struct is *ABI-public* (see [SE-0193][]), i.e. `public` or marked `@abiPublic`.

Reviewer 5: This became `@usableFromInline`, and I forget if we left the "ABI-public" term in SE-0193 when it did.

Author: Yeah that term's used there still.

- If a field has an initial value, the expression computing the initial value does not reference any types or functions that are not ABI-public.

```swift
@frozen
public struct Point {
  public var x: Double
  public var y: Double

  public init(_ x: Double, _ y: Double) {
    self.x = x
    self.y = y
  }
}
```

This affects what changes to the struct's fields affect the ABI of the containing library:

| Change | Normal struct | `@frozen` struct
|---|:---:|:---:
| Adding fields | Allowed | **Affects ABI**
| Reordering fields | Allowed | **Affects ABI**
| Removing ABI-public fields | Affects ABI | Affects ABI
| Removing non-ABI-public fields | Allowed | **Affects ABI**
| Changing the type of an ABI-public field | Affects ABI | Affects ABI
| Changing the type of a non-ABI-public field | Allowed | **Affects ABI**
| Changing a stored instance property to computed | Allowed | **Affects ABI**
| Changing a computed instance property to stored | Allowed | **Affects ABI**
| Changing the access of a non-ABI-public field | Allowed | Allowed
| Marking an `internal` field as `@abiPublic` | Allowed | Allowed
| Changing an `internal` ABI-public field to be `public` | Allowed | Allowed

> > Note: This proposal is implemented already and in use by the standard library, albeit under different names. The command-line flag is `-enable-resilience`; the attribute is `@_fixed_layout` for structs, and `@frozen` for enums.

Reviewer 5: We fixed the flag name already, and `@frozen` is `@_frozen` right now.

  [SE-0193]: https://github.com/apple/swift-evolution/blob/master/proposals/0193-cross-module-inlining-and-specialization.md

#### Guarantees

Marking a struct `@frozen` only guarantees that its stored instance properties won't change. This allows the compiler to perform certain optimizations, like ignoring properties that are never accessed, or eliminating redundant loads from the same instance property. However, it does not provide a handful of other guarantees that a C struct might:

- **It is not guaranteed to be "trivial"** ([in the C++ sense][trivial]). A frozen struct containing a class reference or closure still requires reference-counting when copied and when it goes out of scope.

> - **It is not guaranteed to be stored in registers.** Promoting structs to be stored in registers is considered an optimization in Swift, and as such the compiler is never *required* to do so. In particular, a struct with a `weak` reference in it must always be stored in memory if the compiler cannot prove that the weak reference is unused.

Reviewer 5: This isn't an ABI guarantee, so in hindsight I'm not sure it's worth including. (I'm pretty sure I was the one who added it originally.)

- **It does not necessarily have a known size or alignment.** A generic frozen struct's layout might depend on the generic argument provided at run time.

- **Even concrete instantiations may not have a known size or alignment.** A frozen struct with a field that's a *non-*frozen, has a size that may not be known until run time.

- **It is not guaranteed to use the same layout as a C struct with a similar "shape".** If such a struct is necessary, it should be defined in a C header and imported into Swift.

> - **The fields are not guaranteed to be laid out in declaration order.** The compiler may choose to reorder fields to minimize padding while satisfying alignment requirements, for example, .

Reviewer 5: "for example"?

Author: Ah, the example is preceding, I just failed to trim a comma from a removed clause.

This proposal does not change either of these guarantees.

  [trivial]: https://docs.microsoft.com/en-us/cpp/cpp/trivial-standard-layout-and-pod-types
  [SE-0055]: https://github.com/apple/swift-evolution/blob/master/proposals/0055-optional-unsafe-pointers.md

### `@frozen` on `enum` types

Marking an enum as `@frozen` will also allow the compiler to optimize away runtime calls.

In addition, marking an enum as frozen restores the ability of a library user to exhaustively switch over that enum without an `@unknown default:`, because it guarantees no further cases will be added.

## Naming

`@frozen` was used here, as originally used in [SE-0192][], for both fixed-layout structs and enums. 

- While they share a fixing of their layout going forward, declaring an enum as frozen has additional meaning to _users_ of the library, whereas a frozen struct is an implementation detail of the library.

> - It would be reasonable to use `@frozen` to describe an enum without associated values that promises not to *add* any cases with associated values. This would allow for similar optimization as `@frozen` on a struct, while still not declaring that the enum is frozen.

Reviewer 5: Your find-and-replace mucked up this paragraph.

> - Since this feature is only relevant for libraries with binary compatibility concerns, it may be more useful to tie its syntax to `@available` from the start, as with [SE-0193][]. In that case neither `@frozen` nor `@frozen` would be appropriate.

Reviewer 5: We didn't actually do that in SE-0193, and find-and-replace mucked up this paragraph too.

C++ structs may also contain non-trivial fields. Rather than client code being able to copy them element-by-element, C++ generates a *copy constructor* that must be called whenever a struct value is copied (including when it is passed to a function). Copy constructors may also be written explicitly, in which case they may run arbitrary code; this affects what optimizations a C++ compiler can do. However, if the compiler can *see* that the copy constructor has not been customized, it may be able to optimize copies similarly to Swift.

(This is overly simplified, but sufficient for this comparison.)

### Rust

Rust likewise has a custom layout algorithm for the fields of its structs. While Rust has not put too much effort into having a stable ABI (other than interoperation with C), their current design behaves a lot like C's in practice: layout is known at compile-time, and changing a struct's fields will change the ABI.

> (there is an interesting post from last year about [optimizing the layout of Rust structs, tuples, and enums][rust].)

Reviewer 5: Nitpick: now two years ago

### Other languages

Many other languages allow defining structs, but most of them either don't bother to define a stable ABI or don't allow modifying the fields of a struct once it becomes part of an ABI. Haskell and Ocaml fall into this bucket.

## Source compatibility

This change does not impact source compatibility. It has always been a source-compatible change to modify a struct's non-public stored instance properties, and to add a new public stored instance property, as long as the struct's existing API does not change (including any initializers).

## Effect on ABI stability

Currently, the layout of a public struct is known at compile time in both the defining library and in its clients. For a library concerned about binary compatibility, the layout of a non-fixed-contents struct must not be exposed to clients, since the library may choose to add new stored instance properties that do not fit in that layout in its next release, or even remove existing properties as long as they are not public.

> These considerations should not affect libraries shipped with their clients, including SwiftPM packages. These libraries should always have binary stability mode turned off, indicating that the compiler is free to optimize based on the layout of a struct (because the library won't change).

Reviewer 5: Still says "binary stability mode".

## Effect on Library Evolution

The set of binary-compatible changes to a struct's stored instance properties is described above.

Taking an existing struct and marking it `@frozen` is something we'd like to support without breaking binary compatibility, but there is no design for that yet.

> Removing `@frozen` from a struct is not allowed; this would break any existing clients that rely on the struct's existing layout.

Reviewer 5: Enums need to be discussed here too now, although you can probably lift that from the early versions of SE-0192.

Author: I think "type" works here with this level of detail.

Some ideas for how to prevent library authors from breaking the rules accidentally are discussed in "Compatibility checking" under "Future directions".

## Future directions

### Compatibility checking

Of course, the compiler can't stop a library author from modifying the fields of a fixed-contents struct, even though that will break binary compatibility. We already have two ideas on how we could catch mistakes of this nature:

- A checker that can compare APIs across library versions, using swiftmodule files or similar.

- Encoding the layout of a type in a symbol name. Clients could link against this symbol so that they'd fail to launch if it changes, but even without that an automated system could check the list of exported symbols to make sure nothing was removed.

This is the same kind of checking described in [SE-0192][] "Non-exhaustive Enums".

>   [SE-0192]: https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md

Reviewer 5: It got taken out of SE-0192. :-/

If we wanted to add this later, the way we would probably implement it is by saying that accessors for stored instance properties of `@frozen` structs must be marked `@inlinable`, with all of the implications that has for being able to make changes in the future.

## Alternatives considered

### Annotation syntax

#### Naming

`@frozen` comes from [SE-0192][], where it was originally used as a term for for enums that will not gain new cases.

Other than `@frozen`, most names have revolved around the notion of "fixed" somehow: `@fixedContents`, `@fixedLayout`, or even just `@fixed`. We ultimately chose `@frozen` as there is a strong similarity between this feature for both enums and structs.

`final` was suggested at one point, but the meaning is a little far from that of `final` on classes.

#### Modifier or attribute?

This proposal suggests a new *attribute* for structs, `@fixedContents`; it could also be a modifier `fixedContents` (or `fixed`), implemented as a context-sensitive keyword. Because this annotation only affects the struct's *implementation* rather than anything about its use, an attribute seemed more appropriate.

### Command-line flag naming

> As mentioned above, the current spelling of the `-enable-library-evolution` flag is `-enable-resilience`. The term "resilience" has been an umbrella name for the Swift project's efforts to support evolving an API without breaking binary compatibility, but it isn't well-known outside of the Swift world. It's better to have this flag be more self-explanatory.

Reviewer 5: I'd just cut this whole thing, since we have `-enable-library-evolution` now and Apple has put messaging around it. It's not that we *can't* change the name at this point, but the bar to clear is higher.

Author: eh, they were still alternatives considered, we're just not likely to consider them further...
